File: PHP/199. Binary Tree Right Side View.php
<?php
/*
Given the root of a binary tree, imagine yourself standing on the right side of it, 
    return the values of the nodes you can see ordered from top to bottom.
*/
include './BuildTreeNode.php';
/*
class TreeNode {
    public $val = null;
    public $left = null;
    public $right = null;
    function __construct($val = 0, $left = null, $right = null) {
        $this->val = $val;
        $this->left = $left;
        $this->right = $right;
    }
}     
*/
class Solution {

    /**
     * 199. Binary Tree Right Side View
     * @param TreeNode $root
     * @return Integer[]
     */
    function rightSideView($root) {
        if ($root === null) { return []; }
        $result = [];
        $queue = [$root];
    
        while (!empty($queue)) {
            $size = count($queue);
            for ($i = 0; $i < $size; $i++) {
                $node = array_shift($queue);
    
                // The last node processed at each level is the rightmost one
                if ($i == $size - 1) {
                    $result[] = $node->val;
                }
    
                if ($node->left !== null) { $queue[] = $node->left; }
                if ($node->right !== null) { $queue[] = $node->right; }
            }
        }
    
        return $result;
    }
}

$c = new Solution;
//$root = buildTree([1,2,3,null,5,null,4]); // Expect [1,3,4]
//$root = buildTree([1,2,3,4,null,null,null,5]); // Expect [1,3,4,5]
//$root = buildTree([1,null,3]); // Expect [1,3]
$root = buildTree([]); // Expect []
print_r($c->rightSideView($root));
?>

File: Rust/199_Binary_Tree_Right_Side_View.rs
/*
Given the root of a binary tree, imagine yourself standing on the right side of it, 
    return the values of the nodes you can see ordered from top to bottom.
*/
/* 199. Binary Tree Right Side View */
use std::rc::Rc;
use std::cell::RefCell;
use std::collections::VecDeque;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(
        val: i32, 
        left: Option<Rc<RefCell<TreeNode>>>, 
        right: Option<Rc<RefCell<TreeNode>>>
    ) -> Rc<RefCell<TreeNode>> {
        Rc::new(RefCell::new(TreeNode { val, left, right }))
    }
}
struct Solution;
impl Solution {
    pub fn right_side_view(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        if root.is_none() { return vec![]; }
        let mut result = vec![];
        let mut queue = VecDeque::new();
        queue.push_back(root.unwrap());
        
        while !queue.is_empty() {
            let size = queue.len();
            for i in 0..size {
                let node = queue.pop_front().unwrap();
                let node_ref = node.borrow();
                
                if i == size - 1 {
                    result.push(node_ref.val);
                }
                if let Some(left) = &node_ref.left {
                    queue.push_back(left.clone());
                }
                if let Some(right) = &node_ref.right {
                    queue.push_back(right.clone());
                }
            }
        }
        result
    }
}

fn main() {
    let root = TreeNode::new(1, 
        Some(TreeNode::new(1, None, Some(TreeNode::new(5, None, None)))), 
        Some(TreeNode::new(3, None,
            Some(TreeNode::new(4, None, None))
        ))
    );    // Expect [1, 3, 4]
    let result = Solution::right_side_view(Some(root));
    println!("{:?}", result);
}